### DISCUSSION TITLE
classes/class_audiostreamgenerator

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#268] [User-contributed notes]
# classes/class_audiostreamgenerator

Inherits: AudioStream< Resource< RefCounted< Object An audio stream with utilities for procedural sound generation. Description: AudioStreamGenerator is a type of audio stream that does not play ba...

https://docs.godotengine.org/en/stable/classes/class_audiostreamgenerator.html

**Comment by RichardEllicott** [comment]
this generator has a problem, put the "var phase = 0.0" on the main body so it is remebered between calling "fill_buffer" .. call "fill_buffer" in process if you want, it won't fill the buffer while no frames are avilable.... this will create an unbroken sine sound - RichardEllicott

**Reply by RichardEllicott** [reply]
sorry didn't know we can't edit our comments, please see my tweaked example should help people:
https://gist.github.com/RichardEllicott/9515782f8bbf55748993f30548652581

**Reply by Calinou** [reply]
> sorry didn't know we can't edit our comments, please see my tweaked example should help people:

You can edit comments by going to the GitHub discussion page corresponding to this comment thread (click the comment count to access it), then use the edit button located behind the **...** icon next to each comment).

**Comment by BaconEggsRL** [comment]
My AudioStreamGenerator always clicks or pops at the end of the buffer
Any reason for this?

**Reply by Calinou** [reply]
Make sure you smoothly fade it out at the end (over a period of at least 100 samples). `smoothstep()` can be useful here. Otherwise, you will get a click or pop due to the nature of digital audio (any app that plays audio that suddenly stops will do this). It's why most music players have a short fade in/out sequence when pausing audio playback, even if it's unnoticeable to the human ear.

**Reply by BaconEggsRL** [reply]
Ty, this worked. Also seems like there might be some clipping due to volume being too high, esp. when layering multiple sounds

**Comment by popcar2** [comment]
To save people time in the future: Remember to PAUSE YOUR AUDIO STREAM when you're not playing the sound so that it doesn't keep generating it forever! `$AudioStreamPlayer.stream_paused = true`

**Comment by nxidplease** [comment]
For people who are not "Expert sounds engineers" like myself :sweat_smile:, what exactly `get_frames_available()` return? I mean I'm aware of how audio is represented sample rate is like the "resolution" per second of the sound and the values are the "current air pressure" at the output device. I guess my question is why is this not uniform(some calls to it in process return 0 frames), shouldn't it be "synced" with the physics time? I would expect the frames available to be synced with dt? And I guess the broader question is what happens in the lower level(so maybe I don't understand how audio works exactly)?

**Reply by RichardEllicott** [reply]
it's the framecount that the buffer wants fed, returns an int... your process frame or if you like thread you feed the audio buffer wouldn't just be in sync, it needs to request how many frames the audio buffer can accept... if you get into audio processing you'll be dealing with buffers like this everywhere, you're basicly feeding in a small snippet of sound
